File: src/elixir/compress.exs
import ExUnit.Assertions

compress = fn input ->
  chunk_fun = fn
    char, {"", nil} -> {:cont, {char, 1}}
    char, {char, count} -> {:cont, {char, count + 1}}
    char, {prev, count} -> {:cont, "#{count}#{prev}", {char, 1}}
  end

  after_fun = fn {char, count} ->
    {:cont, "#{count}#{char}", nil}
  end

  input
  |> String.codepoints()
  |> Enum.chunk_while({"", nil}, chunk_fun, after_fun)
  |> Enum.join()
end

assert compress.("AAABBAAC") == "3A2B2A1C"

File: src/lua/compress.lua
local function compress(s)
  local i, j = 1, 1
  local t = {}
  while i and i <= #s do
    local c = s:sub(i,i) -- grab 1st char in sequence
    i, j = s:find(c.."+",i) -- count repetitions
    if i then t[#t+1] = tostring(j-i+1)..c end -- save sequence to table
    i = j + 1
  end
  return table.concat(t)
end

assert(compress("AAABBAAC") == "3A2B2A1C")
